### PR TITLE
[SPARK-13996][SQL] Add more not null attributes for Filter codegen

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

JIRA: https://issues.apache.org/jira/browse/SPARK-13996

Filter codegen finds the attributes not null by checking IsNotNull(a) expression with a condition if child.output.contains(a). However, the current approach to checking it is not comprehensive. We can improve it.

E.g., for this plan:

```
val rdd = sqlContext.sparkContext.makeRDD(Seq(Row(1, "1"), Row(null, "1"), Row(2, "2")))
val schema = new StructType().add("k", IntegerType).add("v", StringType)
val smallDF = sqlContext.createDataFrame(rdd, schema)
val df = smallDF.filter("isnotnull(k + 1)")
```

The code snippet generated without this patch:

```
/* 031 */   protected void processNext() throws java.io.IOException {
/* 032 */     /*** PRODUCE: Filter isnotnull((k#0 + 1)) */
/* 033 */
/* 034 */     /*** PRODUCE: INPUT */
/* 035 */
/* 036 */     while (!shouldStop() && inputadapter_input.hasNext()) {
/* 037 */       InternalRow inputadapter_row = (InternalRow) inputadapter_input.next();
/* 038 */       /*** CONSUME: Filter isnotnull((k#0 + 1)) */
/* 039 */       /* input[0, int] */
/* 040 */       boolean filter_isNull = inputadapter_row.isNullAt(0);
/* 041 */       int filter_value = filter_isNull ? -1 : (inputadapter_row.getInt(0));
/* 042 */
/* 043 */       /* isnotnull((input[0, int] + 1)) */
/* 044 */       /* (input[0, int] + 1) */
/* 045 */       boolean filter_isNull3 = true;
/* 046 */       int filter_value3 = -1;
/* 047 */   
/* 048 */       if (!filter_isNull) {
/* 049 */         filter_isNull3 = false; // resultCode could change nullability.
/* 050 */         filter_value3 = filter_value + 1;
/* 051 */     
/* 052 */       }
/* 053 */       if (!(!(filter_isNull3))) continue;
/* 054 */   
/* 055 */       filter_metricValue.add(1);
```

With this patch:

```
/* 031 */   protected void processNext() throws java.io.IOException {
/* 032 */     /*** PRODUCE: Filter isnotnull((k#0 + 1)) */
/* 033 */
/* 034 */     /*** PRODUCE: INPUT */
/* 035 */
/* 036 */     while (!shouldStop() && inputadapter_input.hasNext()) {
/* 037 */       InternalRow inputadapter_row = (InternalRow) inputadapter_input.next();
/* 038 */       /*** CONSUME: Filter isnotnull((k#0 + 1)) */
/* 039 */       /* input[0, int] */
/* 040 */       boolean filter_isNull = inputadapter_row.isNullAt(0);
/* 041 */       int filter_value = filter_isNull ? -1 : (inputadapter_row.getInt(0));
/* 042 */
/* 043 */       if (filter_isNull) continue;
/* 044 */
/* 045 */       filter_metricValue.add(1);
```
## How was this patch tested?

Existing tests.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala [hunk 1]
```diff
@@ -79,16 +79,16 @@ case class Filter(condition: Expression, child: SparkPlan)
...
   // Split out all the IsNotNulls from condition.
   private val (notNullPreds, otherPreds) = splitConjunctivePredicates(condition).partition {
-    case IsNotNull(a) if child.output.contains(a) => true
+    case IsNotNull(a) if a.references.subsetOf(child.outputSet) => true
     case _ => false
   }
 
   // The columns that will filtered out by `IsNotNull` could be considered as not nullable.
-  private val notNullAttributes = notNullPreds.flatMap(_.references)
+  private val notNullAttributes = notNullPreds.flatMap(_.references).distinct.map(_.exprId)
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala]: The attributes from conditions can have different qualifiers than child.output, which makes later `indexOf` call failed. So here use exprId instead.

**Comment by viirya** [comment]
cc @davies

**Comment by viirya** [comment]
ping @davies Please take a look at this to see if it is good for you. Thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala [hunk 2]
```diff
@@ -79,16 +79,16 @@ case class Filter(condition: Expression, child: SparkPlan)
 
   // Split out all the IsNotNulls from condition.
   private val (notNullPreds, otherPreds) = splitConjunctivePredicates(condition).partition {
-    case IsNotNull(a) if child.output.contains(a) => true
+    case IsNotNull(a) if a.references.subsetOf(child.outputSet) => true
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala]: This is not correct, for example, `IsNotNull(IsNull(a))` does not means `a` can not be `null`.

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala]: ah, you are right. With the `NullIntolerant` trait introduced in #11809, we can solve this easily. Wait for that PR to be ready then we can re-visit this. Thanks!

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
Looks like `StateStoreSuite` a bit flaky...

**Comment by viirya** [comment]
retest this please.

**Comment by davies** [comment]
LGTM

**Comment by davies** [comment]
Merging this into master, thanks!

**Comment by viirya** [comment]
Thanks for reviewing this.

**Pull request closed without merging**